File: packages/node/src/Seguranca/Ataques/AtaqueIndependenciaProposito.js
'use strict';

const ResultadoAtaque = require('../ResultadoAtaque.js');
const SkipAtaqueException = require('../SkipAtaqueException.js');
const { randomBytes, bitsDiferentes } = require('../Util.js');

/**
 * A mesma (key, iv) é usada pra derivar o keystream de dados ('enc') e a
 * chave do MAC ('mac'). Se os dois streams não forem independentes, um
 * atacante que recupere o keystream de dados (ataque de texto conhecido)
 * pode derivar a chave do MAC e FORJAR tokens válidos.
 *
 * O teste procura dois sintomas de acoplamento:
 *  1. a distância de Hamming média entre os streams deve ser ~50%;
 *  2. XOR(enc, mac) NÃO pode se repetir entre (key, iv) diferentes - se for
 *     uma máscara fixa, o mac é 100% previsível a partir do enc.
 */
class AtaqueIndependenciaProposito {
  constructor({ amostras = 2000, tamanho = 32 } = {}) {
    this.amostras = amostras;
    this.tamanho = tamanho;
  }

  nome() {
    return 'Independência entre keystreams de propósitos diferentes (enc x mac)';
  }

  executar(alvo) {
    const primeiro = alvo.gerarKeystreamBruto(
      alvo.chaveDeTeste(),
      randomBytes(alvo.tamanhoIv()),
      'enc',
      this.tamanho,
    );
    if (primeiro == null) {
      throw new SkipAtaqueException('Alvo não expõe gerarKeystreamBruto.');
    }

    const distancias = [];
    const mascaras = new Set();
    for (let i = 0; i < this.amostras; i++) {
      const chave = randomBytes(alvo.chaveDeTeste().length);
      const iv = randomBytes(alvo.tamanhoIv());

      const enc = alvo.gerarKeystreamBruto(chave, iv, 'enc', this.tamanho);
      const mac = alvo.gerarKeystreamBruto(chave, iv, 'mac', this.tamanho);

      distancias.push(bitsDiferentes(enc, mac) / (this.tamanho * 8));
      mascaras.add(this.xorBytes(enc, mac).toString('hex'));
    }

    const media = distancias.reduce((acc, v) => acc + v, 0) / distancias.length;
    const mascarasDistintas = mascaras.size;

    const vulneravel = media < 0.4 || media > 0.6 || mascarasDistintas < this.amostras;

    return new ResultadoAtaque(
      this.nome(),
      vulneravel,
      vulneravel ? 'critica' : 'info',
      `Hamming médio enc x mac=${(media * 100).toFixed(1)}% (esperado ~50%); ${mascarasDistintas} máscara(s) XOR distinta(s) em ${this.amostras} amostras${
        mascarasDistintas < this.amostras
          ? ' - MÁSCARA REPETIDA: mac previsível a partir de enc!'
          : ''
      }`,
      { media, mascaras_distintas: mascarasDistintas },
    );
  }

  xorBytes(a, b) {
    const len = Math.min(a.length, b.length);
    const out = Buffer.alloc(len);
    for (let i = 0; i < len; i++) {
      out[i] = a[i] ^ b[i];
    }
    return out;
  }
}

module.exports = AtaqueIndependenciaProposito;

File: packages/python/src/Seguranca/Ataques/AtaqueIndependenciaProposito.py
"""A mesma (key, iv) é usada pra derivar o keystream de dados ('enc') e a
chave do MAC ('mac'). Se os dois streams não forem independentes, um
atacante que recupere o keystream de dados (ataque de texto conhecido)
pode derivar a chave do MAC e FORJAR tokens válidos.

O teste procura dois sintomas de acoplamento:
 1. a distância de Hamming média entre os streams deve ser ~50%;
 2. XOR(enc, mac) NÃO pode se repetir entre (key, iv) diferentes - se for
    uma máscara fixa, o mac é 100% previsível a partir do enc.
"""

from ..ResultadoAtaque import ResultadoAtaque
from ..SkipAtaqueException import SkipAtaqueException
from ..Util import bits_diferentes, random_bytes


class AtaqueIndependenciaProposito:
    def __init__(self, amostras: int = 2000, tamanho: int = 32):
        self.amostras = amostras
        self.tamanho = tamanho

    def nome(self) -> str:
        return (
            "Independência entre keystreams de propósitos diferentes "
            "(enc x mac)"
        )

    def executar(self, alvo) -> ResultadoAtaque:
        primeiro = alvo.gerar_keystream_bruto(
            alvo.chave_de_teste(), random_bytes(alvo.tamanho_iv()), "enc", self.tamanho
        )
        if primeiro is None:
            raise SkipAtaqueException("Alvo não expõe gerar_keystream_bruto.")

        distancias = []
        mascaras = {}
        for _ in range(self.amostras):
            chave = random_bytes(len(alvo.chave_de_teste().encode()))
            iv = random_bytes(alvo.tamanho_iv())

            enc = alvo.gerar_keystream_bruto(chave, iv, "enc", self.tamanho)
            mac = alvo.gerar_keystream_bruto(chave, iv, "mac", self.tamanho)

            distancias.append(bits_diferentes(enc, mac) / (self.tamanho * 8))
            mascaras[self._xor_bytes(enc, mac)] = True

        media = sum(distancias) / len(distancias)
        mascaras_distintas = len(mascaras)

        vulneravel = (
            media < 0.40 or media > 0.60 or mascaras_distintas < self.amostras
        )

        detalhe = (
            f"Hamming médio enc x mac={media * 100:.1f}% (esperado ~50%); "
            f"{mascaras_distintas} máscara(s) XOR distinta(s) em "
            f"{self.amostras} amostras"
        )
        if mascaras_distintas < self.amostras:
            detalhe += " - MÁSCARA REPETIDA: mac previsível a partir de enc!"

        return ResultadoAtaque(
            self.nome(),
            vulneravel,
            "critica" if vulneravel else "info",
            detalhe,
            {"media": media, "mascaras_distintas": mascaras_distintas},
        )

    def _xor_bytes(self, a: bytes, b: bytes) -> bytes:
        return bytes(x ^ y for x, y in zip(a, b))
